src/main/java/com/github/rholder/retry/Retryer.java

```diff
@@ -84,69 +85,109 @@
      * @return the computed result of the given callable
-     * @throws ExecutionException if the given callable throws an exception, and the
-     *                            rejection predicate considers the attempt as successful.
-     *                            The original exception is wrapped into an ExecutionException.
-     * @throws RetryException     if all the attempts failed before the stop strategy decided
-     *                            to abort, or the thread was interrupted. Note that if the thread
-     *                            is interrupted, this exception is thrown and the thread's
-     *                            interrupt status is set.
+     * @throws Exception            if the given callable throws an exception, and the
+     *                              rejection predicate considers the attempt as successful.
+     * @throws InterruptedException if the current thread is interrupted
+     * @throws RetryException       if all the attempts failed and the last attempt
+     *                              ended with a result rather than an exception.
      */
-    public <T> T call(Callable<T> callable) throws ExecutionException, RetryException {
+    public <T> T call(Callable<T> callable) throws Exception {
         long startTime = System.nanoTime();
         for (int attemptNumber = 1; ; attemptNumber++) {
-            Attempt<T> attempt;
-            try {
-                T result = attemptTimeLimiter.call(callable);
-                long delaySinceFirstAttempt = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTime);
-                attempt = new ResultAttempt<>(result, attemptNumber, delaySinceFirstAttempt);
-            } catch (Throwable t) {
-                long delaySinceFirstAttempt = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTime);
-                attempt = new ExceptionAttempt<>(t, attemptNumber, delaySinceFirstAttempt);
-            }
-
-            for (RetryListener listener : listeners) {
-                listener.onRetry(attempt);
-            }
 
-            if (!test(attempt)) {
+            Attempt<T> attempt = makeAttempt(callable, startTime, attemptNumber);
+            if (!shouldReject(attempt)) {
                 return attempt.get();
             }
+
             if (stopStrategy.shouldStop(attempt)) {
-                throw new RetryException(attemptNumber, attempt);
-            } else {
-                long sleepTime = waitStrategy.computeSleepTime(attempt);
-                try {
-                    blockStrategy.block(sleepTime);
-                } catch (InterruptedException e) {
-                    Thread.currentThread().interrupt();
+                if (!attempt.hasException()) {
                     throw new RetryException(attemptNumber, attempt);
                 }
+                Throwable throwable = attempt.getException();
+                if (throwable instanceof Exception) {
+                    throw (Exception) throwable;
```

Comment:
Isn't this is behavior change for InterruptedException?  I'm not 100% sure, but by catching Throwable in makeAttempt, aren't we responsible for calling Thread.currentThread().interrupt() in case we're really dealing with an InterruptedException?  I've stared at http://www.yegor256.com/2015/10/20/interrupted-exception.html for awhile and I think so, but could be convinced otherwise.

Reply:
You are correct. I was going to add some tests for InterruptedExceptions before merging. I like to think I would have caught it, but maybe not. Thank you.

It makes me want to rethink this approach. The original guava-retrying also wrapped InterruptedExceptions, but it correctly interrupted the threads. However, I'm not sure throwing a non-InterruptedException is a good idea. It seems to me callers are very unlikely to handle it correctly.

I wanted the unchecked methods to make it easier to retry with lambdas that don't throw exceptions. But given that the retryer sleeps and thus can throw Interrupted, maybe it should just propagate InterruptedException. That will make it explicit to the caller. Of course, that means giving up the unchecked methods.

So perhaps I should rip out the unchecked methods again (bummer), then change the signatures of call and run to this:

```java
public <T> T call(Callable<T> callable) throws RetryException, InterruptedException
public void run(Runnable runnable) throws InterruptedException
```

Also, to get rid of the generic "throws Exception" in call, I should return to wrapping any exception that stops the attempt to be wrapped in RetryException. It's disappointing I can't make using lambdas easier, but I don't see any other way that is safe.

@dbyron0, what do you think?

Reply:
I think ditching ExecutionException is good progress.  It would clean up my code.  I think this means the change to what RetryException means sticks around as well.  Being able to get rid of `throws Exception` seems good too.

I haven't totally digested the unchecked methods, but after a quick look I bet I stick with the `call` method.

Sorry if this isn't super helpful.  I bet tests with InterruptedException go a long way to working this out.

Reply:
Thanks. I'll go with the plan outlined above for now.